### PR TITLE
Refactor IndexScanner

### PR DESCRIPTION
**Pull request by gczsjdy**
## What changes were proposed in this pull request?

After this refactoring, it's easier to revisit our deciding polices (using index or not)
I changed from chaotic `if-else` based(policy based) decision -> clear policy based, and maybe future -> cost based

Refactoring:
- Reduce unnecessary funcs, unused fields (tryAnalysis, multiple readBehaviors, canBeOptimzied... etc)
- Rename ambiguous func names like readBehavior
- Make code structure more clear, intuitive. Reduce logics
- Comments
- Remove some logInfo, logDebug, we can provide a debug patch for debugging need of customers instead of the current way

## How was this patch tested?

N/A

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala [hunk 1]
```diff
@@ -71,99 +67,86 @@ private[oap] abstract class IndexScanner(idxMeta: IndexMeta)
...
+
+    val filePolicyEnable =
+      conf.getBoolean(OapConf.OAP_EXECUTOR_INDEX_SELECTION_FILE_POLICY.key,
         OapConf.OAP_EXECUTOR_INDEX_SELECTION_FILE_POLICY.defaultValue.get)
-      if (filePolicyEnable && indexFileSize > dataFileSize * ratio) {
-        return StatsAnalysisResult.FULL_SCAN
-      }
 
+    val indexFileSize = indexPath.getFileSystem(conf).getContentSummary(indexPath).getLength
+    val dataFileSize = dataPath.getFileSystem(conf).getContentSummary(dataPath).getLength
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: fs.getFileStatus is enough and we should assert indexFileSize.isFile & dataFileSize.isFile return true

Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: `fs.getFileStatus().getCapacity()`? it's similar with current way

Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: For `isFile`, the 2 path comes originally from `file: PartitionFile` from`buildReaderWithPartitionValues`, so I guess we don't need this check?

Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: fs.getFileStatus().getlen

Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: indexPath is build by `IndexUtils.indexFileFromDataFile`,What will happen here if indexPath is a dir？

Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: ContentSummary is a class contains filestatus,filecount and dircount, we only need filestatus, use `fs.getFileStatus` api ContentSummary instance will not created.

**Comment by linhong-intel** [comment]
lgtm, the policy looks much more clear now. thanks @gczsjdy

**Comment by gczsjdy** [comment]
Thanks @linhong-intel @LuciferYang !

**Pull request merged**